### PR TITLE
[New] Support shorthand fragment syntax

### LINKED ISSUE #1694: react/prop-types doesn't work with fragment syntax
**Example:**

```javascript
import React from "react";

const ForAttendees = ({ page }) => (
  <>
    <section className="intro intro_attendees">{page}</section>
    <div className="grid--5col">barfoo</div>
  </>
);

export default ForAttendees;
```

`react/prop-types` does **not** capture this. If you replace the fragment syntax with divs, it detects the case.

### PR DESCRIPTION
**Pull request by alexzherdev**
Resolves #1694.

### CODE REVIEW

File: lib/util/jsx.js [hunk 1]
```diff
@@ -25,6 +25,16 @@ function isDOMComponent(node) {
   return COMPAT_TAG_REGEX.test(name);
 }
 
+/**
+ * Checks if a node represents a JSX element or fragment.
+ * @param {object} node - node to check.
+ * @returns {boolean} Whether or not the node if a JSX element or fragment.
+ */
+function isJSX(node) {
```
Comment (alexzherdev) [inline lib/util/jsx.js]: Suggestions for a better name welcome (`isJSXElementOrFragment` seemed too wordy)

File: tests/lib/rules/jsx-curly-brace-presence.js [hunk 2]
```diff
@@ -32,6 +32,10 @@ ruleTester.run('jsx-curly-brace-presence', rule, {
     {
       code: '<App {...props}>foo</App>'
     },
+    {
+      code: '<>foo</>',
```
Comment (alexzherdev) [inline tests/lib/rules/jsx-curly-brace-presence.js]: This rule doesn't really care whether it's an element or a fragment, so I just added one simple valid test, and one invalid.

**Comment by ljharb** [comment]
I'll wait to merge this til we've updated all the relevant rules.

**Comment by alexzherdev** [comment]
Quick q: I noticed some of the rules are using selectors e.g. https://github.com/yannickcr/eslint-plugin-react/blob/master/lib/rules/jsx-curly-brace-presence.js#L249
Selectors were introduced in ESLint 3.18 and we seem to be supporting ^3.0.0, so I guess we should avoid using them?

**Comment by ljharb** [comment]
Fair point; altho nobody's complained. I guess we should fix that.

**Comment by alexzherdev** [comment]
I was asking mostly because this change could use the syntax for `'JSXElement, JSXFragment'`. I will avoid it then and fix existing cases separately afterwards.

**Comment by alexzherdev** [comment]
I believe I covered what needed to be covered, perhaps even redundantly.

**Comment by alexzherdev** [comment]
Added support in 4 more rules, plus some tests in other relevant rules.

**Comment by bensaufley** [comment]
@ljharb what's the timeline for the release that will include this?

**Comment by ljharb** [comment]
[v7.12.0](https://github.com/yannickcr/eslint-plugin-react/releases/tag/v7.12.0) is released.
